odoo/addons/base/tests/test_base.py

```diff
@@ -52,13 +64,275 @@ def test_05_safe_eval_forbiddon(self):
+        )
+
+    def test_08_subscript(self):
+        lst = [
+            0, sys
+        ]
+
+        dico = {
+            "a": "b",
+            "c": sys,
+            sys: lambda _: 1
+        }
+
+        code1 = "lst[0]"
+        code2 = "lst[1]"
+        code3 = "dico['a']"
+        code4 = "dico['c']"
+        code5 = "dico[sys]"
+
+        safe_eval(code1, globals_dict={"lst": lst})
+
+        with self.assertRaises(ValueError):
+            safe_eval(code2, globals_dict={"lst": lst})
+
+        safe_eval(code3, globals_dict={"dico": dico})
+
+        with self.assertRaises(ValueError):
+            safe_eval(code4, globals_dict={"dico": dico})
+
+        with self.assertRaises(ValueError):
+            safe_eval(code5, globals_dict={"dico": dico, "sys": sys})
+
+    def test_09_misc_escape(self):
+        class StrictStr:
+            """
+            This class represents an object that is defined in the code base.
+            An attacker can't modify this object.
+            """
+
+            def __init__(self, val):
+                self.obj_attr = (
+                    "format",
+                    "lower",
+                )
+
+                if not isinstance(val, str):
+                    raise TypeError("Expected a string.")
+
+                self.val = val
+
+            def getattrib(self, name):
+                # For some reason, it's possible to meet those kind of method in the wild
+                if name not in self.obj_attr:
+                    raise AttributeError(f"Attribute '{name}' is not allowed.")
+                return self.obj.__getattribute__(name)
+
+        class WeirdClass(StrictStr):
+            def addAttr(self, name, val):
+                # I hope that nobody will ever do this
+                self.__setattr__(name, val)
+
+        code = """
+        w = WeirdClass()
+        w.addAttr("obj", print)
+        w.addAttr("obj_attr", ('__self__', ))
+        w.getattrib("__self__").exec("import this")
+        """
+
+        with self.assertRaises(ValueError):
+            safe_eval(cleandoc(code), globals_dict={
+                      "StrictStr": StrictStr, "WeirdClass": WeirdClass, "print": print}, mode="exec")
+
+    def test_10_format_should_be_denied(self):
+        # format is forbidden on strings
+        # because it can be used to leak data through dunders
+
+        codefmt = cleandoc(
+            """
+            fmt = "Hello {name}"
+            fmt.format(name="World")
+            """
+        )
+
+        codefmt2 = cleandoc(
+            """
+            fmt = "Hello {}"
+            fmt.format("World")
+            """
+        )
+
+        codefmt3 = cleandoc(
+            """
+            fmt = "Hello {name}"
+            fmt.format_map({"name": "World"})
+            """
+        )
+
+        codeCfmt = cleandoc(
```

Comment:
same inline that bellow, it's a bit annoying to follow the test (same for all tests)